Ejercicio 2(Python): Para el dataset de marketing, realice las pruebas de hipótesis utilizando la tabla de ANOVA sobre: 1) Uno de los parámetros, justificar cual. 2) Todos los parámetros. 3)Un subconjunto de parámetros, justificar cual. Usar la tabla de ANOVA para calcular el estadístico F0 y encontrar el p-valor asociado usando pf(F_0, df1, df2, lower.tail = FALSE). Justifique su respuesta y suba su código en Python a github.

In [3]:
import pandas as pd
import statsmodels.formula.api as smf

# Leer el archivo CSV generado en R
marketing = pd.read_csv("marketing_dataset.csv")

# Mostrar estructura
print("Estructura del dataset:")
print(marketing.info())
print("\nPrimeras filas:")
print(marketing.head())

Estructura del dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   youtube    200 non-null    float64
 1   facebook   200 non-null    float64
 2   newspaper  200 non-null    float64
 3   sales      200 non-null    float64
dtypes: float64(4)
memory usage: 6.4 KB
None

Primeras filas:
   youtube  facebook  newspaper  sales
0   276.12     45.36      83.04  26.52
1    53.40     47.16      54.12  12.48
2    20.64     55.08      83.16  11.16
3   181.80     49.56      70.20  22.20
4   216.96     12.96      70.08  15.48


In [9]:
modeloF = smf.ols(formula='sales ~ facebook', data=marketing).fit()
modeloY = smf.ols(formula='sales ~ youtube', data=marketing).fit()
modeloN = smf.ols(formula='sales ~ newspaper', data=marketing).fit()

# Obtener los R² de cada modelo
r2_facebook = modeloF.rsquared
r2_youtube = modeloY.rsquared
r2_newspaper = modeloN.rsquared

# Imprimir los resultados (equivalente a cat() en R)
print(f"R2 facebook: {r2_facebook}")
print(f"R2 youtube: {r2_youtube}")
print(f"R2 newspaper: {r2_newspaper}")


R2 facebook: 0.33203245544529536
R2 youtube: 0.611875050850071
R2 newspaper: 0.05212044544430494


RESPUESTA  1) Uno de los parámetros, justificar cual.
Resultados del codigo anterior:

R2 facebook: 0.33203245544529536
R2 youtube: 0.611875050850071
R2 newspaper: 0.05212044544430494

El modelo con sólo youtube es el que explica mejor los datos R^2 = 0.611.

2) Todos los parámetros.

In [14]:
modelo_completo = smf.ols(formula='sales ~ youtube + facebook + newspaper', data=marketing).fit()

print("R2 modelo completo:", modelo_completo.rsquared)
print(modelo_completo.summary())

R2 modelo completo: 0.8972106381789522
                            OLS Regression Results                            
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Fri, 12 Sep 2025   Prob (F-statistic):           1.58e-96
Time:                        20:29:11   Log-Likelihood:                -422.65
No. Observations:                 200   AIC:                             853.3
Df Residuals:                     196   BIC:                             866.5
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    

Del resultado anterior tenemos: 

R2 modelo completo: 0.8972106381789522
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      3.5267      0.374      9.422      0.000       2.789       4.265
youtube        0.0458      0.001     32.809      0.000       0.043       0.049
facebook       0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================

El modelo completo tiene R2 = 0.89721 (mayor que modelo solo con youtube).
De los valores de p-value por parámetro, vemos que el parámetro newspaper tiene un p-value=0.86,
lo cual indica que no debería estar incluido en el modelo. 

3) Un subconjunto de parámetros, justificar cual.

Del análisis anterior, el subconjunto debe contener a youtube y facebook.

In [18]:
import statsmodels.api as sm

modeloY = smf.ols(formula='sales ~ youtube', data=marketing).fit()
modeloFY = smf.ols(formula='sales ~ facebook + youtube', data=marketing).fit()

resultado_anova = sm.stats.anova_lm(modeloY, modeloFY)
print(resultado_anova)
print("R2 modeloFY:", modeloFY.rsquared)
print("R2_adj modeloFY:", modeloFY.rsquared_adj)
print("R2_adj modelo_completo:", modelo_completo.rsquared_adj)


   df_resid          ssr  df_diff      ss_diff           F        Pr(>F)
0     198.0  3027.644040      0.0          NaN         NaN           NaN
1     197.0   801.956131      1.0  2225.687908  546.738781  9.776972e-59
R2 modeloFY: 0.8971942610828956
R2_adj modeloFY: 0.8961505479974428
R2_adj modelo_completo: 0.8956373316204668


De los resultados obtenidos:

df_resid          ssr  df_diff      ss_diff           F        Pr(>F)
0     198.0  3027.644040      0.0          NaN         NaN           NaN
1     197.0   801.956131      1.0  2225.687908  546.738781  9.776972e-59
R2 modeloFY: 0.8971942610828956
R2_adj modeloFY: 0.8961505479974428
R2_adj modelo_completo: 0.8956373316204668

Se ve que la reducción de error es de 3027.644 (con youtube) a 801.956 (con youtube y facebook) y
el valor de p-value es casi 0, lo cual indica que el modelo es mucho mejor. 
El valor de R^2 es de 0.89719, casi igual al del modelo completo (de 0.89721).

El modeloFY es mejor que modelo_completo ya que R^2_adj es mas alto (0.8961) que el modelo completo (0.8956).